crates/oxc_linter/src/rules/unicorn/prefer_reflect_apply.rs

```diff
@@ -0,0 +1,162 @@
+use oxc_ast::{
+    ast::{Argument, Expression, MemberExpression},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error(
+    "eslint-plugin-unicorn(prefer-reflect-apply): Prefer Reflect.apply() over Function#apply()"
+)]
+#[diagnostic(severity(warning), help("Reflect.apply() is less verbose and easier to understand."))]
+struct PreferReflectApplyDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct PreferReflectApply;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Reflect.apply() is arguably less verbose and easier to understand.
+    /// In addition, when you accept arbitrary methods,
+    /// it's not safe to assume .apply() exists or is not overridden.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// foo.apply(null, [42]);
+    ///
+    /// // Good
+    /// Reflect.apply(foo, null);
+    /// ```
+    PreferReflectApply,
+    style
+);
+
+fn is_apply_signature(argument_1: &Argument, argument_2: &Argument) -> bool {
+    let Argument::Expression(arg_expr_1) = argument_1 else {
+        return false;
+    };
+
+    let Argument::Expression(arg_expr_2) = argument_2 else {
+        return false;
+    };
+
+    if (matches!(arg_expr_1, Expression::ThisExpression(_))
+        || matches!(arg_expr_1, Expression::NullLiteral(_)))
+    {
+        if matches!(arg_expr_2, Expression::ArrayExpression(_)) {
+            return true;
+        }
+
+        if let Expression::Identifier(iden) = arg_expr_2 {
+            return iden.name == "arguments";
+        }
+    }
+
+    false
+}
+
+fn is_static_property_name_equal(expr: &MemberExpression, value: &str) -> bool {
+    expr.static_property_name().is_some_and(|name| name == value)
+}
+
+impl Rule for PreferReflectApply {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+
+        let Expression::MemberExpression(member_expr) = &call_expr.callee else {
+            return;
+        };
+
+        if call_expr.optional
+            || matches!(member_expr.object(), Expression::ArrayExpression(_))
+            || matches!(member_expr.object(), Expression::ObjectExpression(_))
+            || member_expr.object().is_literal()
+        {
+            return;
+        }
+
+        if is_static_property_name_equal(member_expr, "apply")
+            && call_expr.arguments.len() == 2
+            && is_apply_signature(&call_expr.arguments[0], &call_expr.arguments[1])
+        {
+            ctx.diagnostic(PreferReflectApplyDiagnostic(call_expr.span));
+            return;
+        }
+
+        if is_static_property_name_equal(member_expr, "call") {
```

Comment:
NIT: 
```rs
        if is_static_property_name_equal(member_expr, "call") {
            let Expression::MemberExpression(member_expr_obj) = member_expr.object() else {
                return;
            };
            if is_static_property_name_equal(member_expr_obj, "apply") {
                let Expression::MemberExpression(member_expr_obj_obj) = member_expr_obj.object()
                else {
                    return;
                };
                if is_static_property_name_equal(member_expr_obj_obj, "prototype") {
                    let Expression::Identifier(iden) = member_expr_obj_obj.object() else { return };
                    if iden.name == "Function"
                        && matches!(call_expr.arguments.as_slice(), [_, second, third] if is_apply_signature(&second, &third))
                    {
                        ctx.diagnostic(PreferReflectApplyDiagnostic(call_expr.span));
                    }
                }
            }
        }

```

Reply:
Thanks for your advice. I've made some adjustments.